### ISSUE TITLE
Connections and objects belonging to connections are not fully garbage collected (memory leak)

### ISSUE THREAD

**Issue by ibc** [eventmachine/eventmachine#157]
It seems that when a TCP connection has been closed (and it's not referenced by alive Ruby objects) such connection and objects belonging to it remain active in memory and **are not** garbage collected until a new request(s) arrive in a new connection, so the remainin connection/objets are the last ones (not entirely true, see below).

In case of using threads (EM.defer) then as many connections (and their objects) can remain in memory (I suspect they are referenced somehow in the thread which processed the request message).

I provide a client and server to demostrate it (some code as in bug 157 but here with other purpose):
  https://gist.github.com/781286

_NOTE:_ I've done some changes to the gist, please download the clietn and server again.

The server works as follows:
- It listens in port 6666.
- When it receives "A" it replies "a" as follows:
  send_data "a"
- When it receives "B" it replies "b" as follows:
  EM.next_tick do
    send_data "b"
  end
- When it receives "C" it replies "c" as follows:
  msg.connection = self
  operation = proc { msg }
  callback = proc do |result|
    result.connection.send_data "c"
  end
  EM.defer( operation, callback )
- A EM timer invokes GC.start every 3 seconds and prints in the screen the number of server connection objects (TestLeakServer < EM::Connection) and messages (Message) existing in memory. These objects are not garbage collected. Such objects are retrieved using ObjectSpace module.

The client works as follows:

```
USAGE:  ./em_client_test_leak.rb DATA TIMES CONCURRENCY
  DATA         must be "A" or "B" or "C"
  TIMES        number of messages to send (default 1)
  CONCURRENCY  messages to send together upon receipt of a response (default 1)
```
# Test cases:
## 1) No client connections
- Start the server. After 3 seconds you can see:
  GC.start executed
  ObjectSpace.each_object:
  TOTAL objets:              13456
  TOTAL EM::Connection:      0
  TOTAL TestLeakServer:      0
  TOTAL Message:             0

Of course it's ok.
## 2) Single request and no EM threads
- Run the client as follows:
  ./em_client_test_leak.rb A 1
- The TCP connection is closed by the client upon receipt of the response. But after some second you can see in the server:
  GC.start executed
  ObjectSpace.each_object:
  TOTAL objets:              13462
  TOTAL EM::Connection:      1
  TOTAL TestLeakServer:      1
  - connection: source port = 54121 | error? = true | object_id = 15593400
    TOTAL Message:             1
  - message source port: 54121

Note that the already closed connection (source port 54121) remains in memory (but #error? says true), and also the Message created over it.
- Run the client again:
  ./em_client_test_leak.rb A 1
- Now in server screen:
  GC.start executed
  ObjectSpace.each_object:
  TOTAL objets:              13462
  TOTAL EM::Connection:      1
  TOTAL TestLeakServer:      1
  - connection: source port = 54129 | error? = true | object_id = 16393300
    TOTAL Message:             1
  - message source port: 54129

Now the last client connection is the remaining one!
## 3) Single request with EM threads
- Restart the server. Note that EM.threadpool_size = 10.
- Run the client as follows:
  ./em_client_test_leak.rb C 1
- In the server you get:
  GC.start executed
  ObjectSpace.each_object:
  TOTAL objets:              13507
  TOTAL EM::Connection:      1
  TOTAL TestLeakServer:      1
  - connection: source port = 56413 | error? = true | object_id = 20657660
    TOTAL Message:             1
  - message source port: 56413
- Run the same command multiples times (at least 20), finally you would get this:
  GC.start executed
  ObjectSpace.each_object:
  TOTAL objets:              13588
  TOTAL EM::Connection:      10
  TOTAL TestLeakServer:      10
  - connection: source port = 56465 | error? = true | object_id = 21450140
  - connection: source port = 56464 | error? = true | object_id = 21451480
  - connection: source port = 56463 | error? = true | object_id = 21452560
  - connection: source port = 56462 | error? = true | object_id = 21453640
  - connection: source port = 56468 | error? = true | object_id = 21453740
  - connection: source port = 56467 | error? = true | object_id = 21454620
  - connection: source port = 56461 | error? = true | object_id = 21454640
  - connection: source port = 56466 | error? = true | object_id = 21455460
  - connection: source port = 56460 | error? = true | object_id = 21455580
  - connection: source port = 56459 | error? = true | object_id = 21456820
    TOTAL Message:             10
  - message source port: 56465
  - message source port: 56464
  - message source port: 56463
  - message source port: 56462
  - message source port: 56468
  - message source port: 56461
  - message source port: 56467
  - message source port: 56466
  - message source port: 56460
  - message source port: 56459

This means that each thread is storing, somehow, a dead connection object and the Message created over it. Of course the number of remaining Ruby objects has been incremented (TOTAL objets).
## 4) Multiple requests together in same TCP stream with EM threads => memory leak
- Restart the server as follows (just SpaceObjects info is printed in the screen):
  ./em_server_test_leak.rb >/dev/null
- Run the client (send 1000 requests after each received response, stop when sending 1000000 requests):
  /em_client_test_leak.rb C 1000000 1000

Run a top in other terminal and you can see how the EM server memory usage grows up **very fast**!!! And when the client terminates (the connection is closed) the memory is **not** freed.

However I don't understand which data is filling so many memory as ObjectSpace statics just show ~10 remaining connections and Message objects.
- Now in client run:
  ./em_client_test_leak.rb C 1

You can see in top command that EM server has freed "some" memory (but not all).

Basically this means that a single client over a single connection could overload the server (a very easy DoS attack).
